src/lib.rs

```diff
@@ -157,7 +157,35 @@ impl Deref for SmolStr {
 
 impl PartialEq<SmolStr> for SmolStr {
     fn eq(&self, other: &SmolStr) -> bool {
-        self.as_str() == other.as_str()
+        match (&self.0, &other.0) {
```

Comment:
This requires a canonical representation (see https://github.com/rust-analyzer/smol_str/pull/62/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R87).